### PR TITLE
feat: workflow executor

### PR DESCRIPTION
**Pull request by darkskygit**
fix AFF-1221 AFF-1232

### CODE REVIEW

**Comment by darkskygit** [comment]
* **#7311** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7311?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7231** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7231?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7159** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7159?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#7123** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7123?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7119** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7119?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @darkskygit and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/backend/server/src/plugins/copilot/workflow/node.ts [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    return this.edges.length;
+  }
+
+  private async evaluateCondition(
+    condition: string,
+    params: WorkflowNodeState
+  ): Promise<string | undefined> {
+    try {
+      const chooseNode = this.worker.run({
+        iife: `(${condition})(nodeIds, params)`,
```
Comment (forehalo) [inline packages/backend/server/src/plugins/copilot/workflow/node.ts]: why do we need to send condition to a worker to execute? the conditions are programmatically executable and predefined, all args are known at the mean time

Comment (darkskygit) [reply packages/backend/server/src/plugins/copilot/workflow/node.ts]: we will allow users to modify prompts in the admin dashboard in future, then the workflow graph should also allow console modifications
in this case, the condition will not only be predefined, but may also be unknown risk inputs
so we need to use workers to isolate untrusted code

Comment (forehalo) [reply packages/backend/server/src/plugins/copilot/workflow/node.ts]: It must be a long story allowing users modifications, and I prefer to use worker when it's really needed

Comment (darkskygit) [reply packages/backend/server/src/plugins/copilot/workflow/node.ts]: switched to evaluate via new Function in unmodified state

File: packages/backend/server/src/plugins/copilot/workflow/node.ts [hunk 2]
```diff
@@ -1,21 +1,70 @@
...
+    private readonly data: NodeData
+  ) {
+    if (data.nodeType === WorkflowNodeType.Basic) {
+      this.executor = getWorkflowExecutor(data.type);
+    } else if (data.nodeType === WorkflowNodeType.Decision) {
+      // prepare decision condition, reused in each run
+      const iife = `(${data.condition})(nodeIds, params)`;
+      // only eval the condition in worker if graph has been modified
+      if (graph.modified) {
+        const worker = new Piscina({
```
Comment (forehalo) [inline packages/backend/server/src/plugins/copilot/workflow/node.ts:37]: still using worker?

Comment (darkskygit) [reply packages/backend/server/src/plugins/copilot/workflow/node.ts:37]: only use worker when modified=true
currently, no workflow sets this flag, and it is reserved and will allow admin to modify it at runtime in the future

File: packages/backend/server/src/plugins/copilot/workflow/node.ts [hunk 3]
```diff
@@ -1,21 +1,70 @@
...
+          execArgv: [],
+        });
+        this.condition = (params: WorkflowNodeState) =>
+          worker.run({
+            iife,
+            nodeIds: this.edges.map(node => node.id),
+            params,
+          });
+      } else {
+        const func = new Function(
```
Comment (forehalo) [inline packages/backend/server/src/plugins/copilot/workflow/node.ts]: make data.condition type `fn | string`? that would be more safe and efficient when not modified cause I don't trust `fn.toString()` much

**Review by forehalo — APPROVED** [review]

**Pull request merged**